avocado/utils/asset.py

```diff
@@ -159,6 +180,11 @@ def fetch(self):
             else:
                 raise UnsupportedProtocolError("Unsupported protocol"
                                                ": %s" % urlobj.scheme)
+            cache_relative_dir = self._get_relative_dir(urlobj)
```

Comment:
OK, you chose to always update the relative dir based on url. Makes sense, but in that case you should not check only for the base location presence but you should iterate through all of them first and only when not found you should start downloading them.

Use case that demonstrates this could be bob and alice serving the same file but bob's one is corrupted. This implementation will:

```python
fetch("foo", ["bob/foo", "alice/foo"], sha=alice_sha)
# downloads bob/foo, deletes bob due to sha
# downloads alice/foo
fetch("foo", ["bob/foo", "alice/foo"], sha=alice_sha)
# downloads bob/foo, deletes bob due to sha
# uses downloaded alice/foo
fetch("foo", ["bob/foo", "alice/foo"], sha=alice_sha)
# downloads bob/foo, deletes bob due to sha
# uses downloaded alice/foo
```

But if we check all possible locations, it'd only download bob/foo and alice/foo once:

```python
fetch("foo", ["bob/foo", "alice/foo"], sha=alice_sha)
# downloads bob/foo, deletes bob due to sha
# downloads alice/foo
fetch("foo", ["bob/foo", "alice/foo"], sha=alice_sha)
# uses downloaded alice/foo
fetch("foo", ["bob/foo", "alice/foo"], sha=alice_sha)
# uses downloaded alice/foo
```

Reply:
Btw this change of behavior has one cons and that is when user wants any netperf. Previously something like:

```
fetch("github/netperf")
fetch("sourceforge/netperf")
```
the second fetch would simply reuse the netperf from github. On the other side I think the benefits are greater and people should learn to put `locations=["github/netperf", "sourceforge/netperf"]` there, I'm mentioning it here record we considered the pros and cons.

Reply:
Sure, but this is also the core reason of the problem we're fixing.  For using the same netperf, a hash would be necessary.

Reply:
sure, it's always about balancing it...